File: src/test/java/org/zenbot/szolnok/timetable/batch/step/bus/processor/JsoupDocumentToTimetableProcessorTest.java
package org.zenbot.szolnok.timetable.batch.step.bus.processor;

import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.zenbot.szolnok.timetable.configuration.properties.TimetableProperties;
import org.zenbot.szolnok.timetable.configuration.properties.TimetableSelectorProperties;
import org.zenbot.szolnok.timetable.domain.Timetable;

import java.util.HashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

public class JsoupDocumentToTimetableProcessorTest {

    private static final String ROUTE_NAME_SELECTOR = "ROUTE_NAME_SELECTOR";
    private static final String FROM_SELECTOR = "FROM_SELECTOR";
    private static final String ACTUAL_STOP_SELECTOR = "ACTUAL_STOP_SELECTOR";
    private static final String ROUTE_NAME = "ROUTE_NAME";
    private static final String START_BUS_STOP = "START_BUS_STOP";
    private static final String END_BUS_STOP = "END_BUS_STOP";
    private static final String ACTUAL_STOP = "ACTUAL_STOP";

    private JsoupDocumentToTimetableProcessor testSubject;

    private TimetableSelectorProperties selectorProperties;

    @Mock
    private Document html;

    @Mock
    private Elements routeNameElement;

    @Mock
    private Elements startBusStopElement;

    @Mock
    private StartBusStopSelectorItemProcessorHelper startBusStopSelectorItemProcessorHelper;

    @Mock
    private EndBusStopSelectorItemProcessorHelper endBusStopSelectorItemProcessorHelper;

    @Mock
    private ActualStopSelectorItemProcessorHelper actualStopSelectorItemProcessorHelper;

    @Mock
    private TimetableRowBuilderItemProcessorHelper timetableRowBuilderItemProcessorHelper;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        selectorProperties = new TimetableSelectorProperties();
        selectorProperties.setRouteNameSelector(ROUTE_NAME_SELECTOR);
        selectorProperties.setFromSelector(FROM_SELECTOR);
        selectorProperties.setActualStopSelector(ACTUAL_STOP_SELECTOR);
        TimetableProperties properties = new TimetableProperties();
        properties.setSelector(selectorProperties);

        testSubject = new JsoupDocumentToTimetableProcessor(properties, startBusStopSelectorItemProcessorHelper, endBusStopSelectorItemProcessorHelper, actualStopSelectorItemProcessorHelper, timetableRowBuilderItemProcessorHelper);
    }

    @Test
    public void processShouldDelegateCallsToHelpers() {
        // GIVEN
        given(html.select(selectorProperties.getRouteNameSelector())).willReturn(routeNameElement);
        given(routeNameElement.text()).willReturn(ROUTE_NAME);
        given(startBusStopSelectorItemProcessorHelper.getStartBusStop(html, selectorProperties)).willReturn(START_BUS_STOP);
        given(endBusStopSelectorItemProcessorHelper.getEndBusStop(any(Document.class), any(TimetableSelectorProperties.class))).willReturn(END_BUS_STOP);
        given(actualStopSelectorItemProcessorHelper.getActualStop(any(Document.class), any(TimetableSelectorProperties.class))).willReturn(ACTUAL_STOP);
        Map<Integer, Map<String, String>> timetable = new HashMap<>();
        given(timetableRowBuilderItemProcessorHelper.getTimetableRows(any(Document.class), any(TimetableSelectorProperties.class))).willReturn(timetable);

        // WHEN
        Timetable result = testSubject.process(html);

        //THEN
        verify(html).select(ROUTE_NAME_SELECTOR);
        verify(startBusStopSelectorItemProcessorHelper).getStartBusStop(html, selectorProperties);
        verify(endBusStopSelectorItemProcessorHelper).getEndBusStop(html, selectorProperties);
        verify(actualStopSelectorItemProcessorHelper).getActualStop(html, selectorProperties);
        verify(timetableRowBuilderItemProcessorHelper).getTimetableRows(html, selectorProperties);

        assertThat(result).isNotNull();
        assertThat(result.getBusName()).isEqualTo(ROUTE_NAME);
        assertThat(result.getStartBusStopName()).isEqualTo(START_BUS_STOP);
        assertThat(result.getActiveStopName()).isEqualTo(ACTUAL_STOP);
        assertThat(result.getEndBusStopName()).isEqualTo(END_BUS_STOP);
        assertThat(result.getTimetable()).isEqualTo(timetable);

    }

}

File: szolnok-timetable-backend-batch-utils/src/test/kotlin/org/zenbot/szolnok/timetable/backend/batch/utils/common/batch/processor/JsoupDocumentToTimetableProcessorTest.kt
package org.zenbot.szolnok.timetable.backend.batch.utils.common.batch.processor

import org.assertj.core.api.Assertions.assertThat
import org.jsoup.nodes.Document
import org.junit.Before
import org.junit.Test
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import org.zenbot.szolnok.timetable.backend.batch.utils.common.batch.processor.helper.ActualStopSelectorItemProcessorHelper
import org.zenbot.szolnok.timetable.backend.batch.utils.common.batch.processor.helper.BusNameSelectorItemProcessorHelper
import org.zenbot.szolnok.timetable.backend.batch.utils.common.batch.processor.helper.StartBusStopSelectorItemProcessorHelper
import org.zenbot.szolnok.timetable.backend.batch.utils.common.batch.processor.helper.TerminalSelectorItemProcessorHelper
import org.zenbot.szolnok.timetable.backend.batch.utils.common.batch.processor.helper.TimetableRowBuilderItemProcessorHelper
import org.zenbot.szolnok.timetable.backend.batch.utils.common.properties.TimetableProperties
import org.zenbot.szolnok.timetable.backend.batch.utils.common.properties.TimetableSelectorProperties

class JsoupDocumentToTimetableProcessorTest {

    private lateinit var testSubject: JsoupDocumentToTimetableProcessor

    private lateinit var properties: TimetableProperties

    @Mock
    private lateinit var busNameSelectorItemProcessorHelper: BusNameSelectorItemProcessorHelper

    @Mock
    private lateinit var startBusStopSelectorItemProcessorHelper: StartBusStopSelectorItemProcessorHelper

    @Mock
    private lateinit var terminalSelectorItemProcessorHelper: TerminalSelectorItemProcessorHelper

    @Mock
    private lateinit var actualStopSelectorItemProcessorHelper: ActualStopSelectorItemProcessorHelper

    @Mock
    private lateinit var timetableRowBuilderItemProcessorHelper: TimetableRowBuilderItemProcessorHelper

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        properties = TimetableProperties()
        testSubject = JsoupDocumentToTimetableProcessor(
                properties,
                busNameSelectorItemProcessorHelper,
                startBusStopSelectorItemProcessorHelper,
                terminalSelectorItemProcessorHelper,
                actualStopSelectorItemProcessorHelper,
                timetableRowBuilderItemProcessorHelper
        )
    }

    @Test
    fun `process should use it's dependencies to create the timetable`() {
        // GIVEN
        val htmlDocument = mock(Document::class.java)
        val timetableMap = mapOf(1 to mapOf("weekday" to "1,2"))
        val selector = TimetableSelectorProperties()
        properties.selector = selector

        given(busNameSelectorItemProcessorHelper.getBusName(htmlDocument, properties.selector)).willReturn("busName")
        given(startBusStopSelectorItemProcessorHelper.getStartBusStop(htmlDocument, properties.selector)).willReturn("startStop")
        given(terminalSelectorItemProcessorHelper.getTerminal(htmlDocument, properties.selector)).willReturn("terminal")
        given(actualStopSelectorItemProcessorHelper.getActualStop(htmlDocument, properties.selector)).willReturn("actual")
        given(timetableRowBuilderItemProcessorHelper.getTimetableRows(htmlDocument, properties.selector)).willReturn(timetableMap)

        // WHEN
        val result = testSubject.process(htmlDocument)

        // THEN
        verify(busNameSelectorItemProcessorHelper).getBusName(htmlDocument, properties.selector)
        verify(startBusStopSelectorItemProcessorHelper).getStartBusStop(htmlDocument, properties.selector)
        verify(terminalSelectorItemProcessorHelper).getTerminal(htmlDocument, properties.selector)
        verify(actualStopSelectorItemProcessorHelper).getActualStop(htmlDocument, properties.selector)
        verify(timetableRowBuilderItemProcessorHelper).getTimetableRows(htmlDocument, properties.selector)

        assertThat(result).isNotNull()
                .hasFieldOrPropertyWithValue("busName", "busName")
                .hasFieldOrPropertyWithValue("startBusStopName", "startStop")
                .hasFieldOrPropertyWithValue("endBusStopName", "terminal")
                .hasFieldOrPropertyWithValue("activeStopName", "actual")
                .hasFieldOrPropertyWithValue("timetable", mutableMapOf(1 to mapOf("weekday" to "1,2")))
    }

}